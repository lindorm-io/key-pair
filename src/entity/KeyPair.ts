import Joi from "joi";
import { Algorithm, KeyPairEvent, KeyType, NamedCurve } from "../enum";
import { IJoseData, IJwk, IKeyJwk } from "../types";
import { JOI_KEY_ALGORITHM, JOI_KEY_ALGORITHMS, JOI_KEY_NAMED_CURVE, JOI_KEY_TYPE } from "../constant";
import { decodeKeys, encodeKeys } from "../util";
import { includes, orderBy } from "lodash";
import {
  EntityBase,
  EntityCreationError,
  IEntity,
  IEntityAttributes,
  IEntityOptions,
  JOI_ENTITY_BASE,
} from "@lindorm-io/entity";

export interface IKeyPairAttributes extends IEntityAttributes {
  algorithms: Array<Algorithm>;
  allowed: boolean;
  expires: Date | null;
  namedCurve: NamedCurve | null;
  passphrase: string | null;
  preferredAlgorithm: Algorithm;
  privateKey: string | null;
  publicKey: string;
  type: KeyType;
}

export interface IKeyPairOptions extends IEntityOptions {
  algorithms: Array<Algorithm>;
  allowed?: boolean;
  expires?: Date;
  namedCurve?: NamedCurve;
  passphrase?: string;
  preferredAlgorithm?: Algorithm;
  privateKey?: string;
  publicKey: string;
  type: KeyType;
}

const schema = Joi.object({
  ...JOI_ENTITY_BASE,

  algorithms: JOI_KEY_ALGORITHMS.required(),
  allowed: Joi.boolean().required(),
  expires: Joi.date().allow(null).required(),
  namedCurve: JOI_KEY_NAMED_CURVE.allow(null).required(),
  passphrase: Joi.string().allow(null).required(),
  preferredAlgorithm: JOI_KEY_ALGORITHM.required(),
  privateKey: Joi.string().allow(null).required(),
  publicKey: Joi.string().required(),
  type: JOI_KEY_TYPE.required(),
});

export class KeyPair extends EntityBase<IKeyPairAttributes> implements IEntity<IKeyPairAttributes> {
  public readonly algorithms: Array<Algorithm>;
  public readonly namedCurve: NamedCurve | null;
  public readonly passphrase: string | null;
  public readonly privateKey: string | null;
  public readonly publicKey: string;
  public readonly type: KeyType;
  private _allowed: boolean;
  private _expires: Date | null;
  private _preferredAlgorithm: Algorithm;

  public constructor(options: IKeyPairOptions) {
    super(options);

    this._allowed = options.allowed !== false;
    this._expires = options.expires || null;
    this._preferredAlgorithm = options.preferredAlgorithm || orderBy(options.algorithms, [(item) => item], ["desc"])[0];

    this.algorithms = options.algorithms;
    this.namedCurve = options.namedCurve || null;
    this.passphrase = options.passphrase || null;
    this.privateKey = options.privateKey || null;
    this.publicKey = options.publicKey;
    this.type = options.type;
  }

  public get allowed(): boolean {
    return this._allowed;
  }
  public set allowed(allowed: boolean) {
    this._allowed = allowed;
    this.addEvent(KeyPairEvent.ALLOWED_CHANGED, { allowed: this._allowed });
  }

  public get expires(): Date | null {
    return this._expires;
  }
  public set expires(expires: Date | null) {
    this._expires = expires;
    this.addEvent(KeyPairEvent.EXPIRES_CHANGED, { expires: this._expires });
  }

  public get preferredAlgorithm(): Algorithm {
    return this._preferredAlgorithm;
  }
  public set preferredAlgorithm(preferredAlgorithm: Algorithm) {
    if (!includes(this.algorithms, preferredAlgorithm)) {
      throw new Error("Invalid preferredAlgorithm");
    }

    this._preferredAlgorithm = preferredAlgorithm;
    this.addEvent(KeyPairEvent.PREFERRED_ALGORITHM_CHANGED, { preferredAlgorithm: this._preferredAlgorithm });
  }

  public create(): void {
    for (const evt of this.events) {
      if (evt.name !== KeyPairEvent.CREATED) continue;
      throw new EntityCreationError(this.constructor.name);
    }

    this.addEvent(KeyPairEvent.CREATED, this.toJSON());
  }

  public getKey(): string {
    return this.id;
  }

  public async schemaValidation(): Promise<void> {
    await schema.validateAsync(this.toJSON());
  }

  public toJSON(): IKeyPairAttributes {
    return {
      ...this.defaultJSON(),

      algorithms: this.algorithms,
      allowed: this.allowed,
      expires: this.expires,
      namedCurve: this.namedCurve,
      passphrase: this.passphrase,
      preferredAlgorithm: this.preferredAlgorithm,
      privateKey: this.privateKey,
      publicKey: this.publicKey,
      type: this.type,
    };
  }

  public toJWK(exposePrivateKey = false): IJwk {
    const data: IKeyJwk = encodeKeys({
      exposePrivateKey,
      namedCurve: this.namedCurve,
      privateKey: this.privateKey,
      publicKey: this.publicKey,
      type: this.type,
    });

    return {
      alg: this.preferredAlgorithm,
      created: this.created.getTime(),
      crv: this.namedCurve ? this.namedCurve : undefined,
      expires: this.expires ? this.expires.getTime() : undefined,
      key_ops: ["sign", "verify"],
      kid: this.id,
      kty: this.type,
      use: "sig",
      ...data,
    };
  }

  public static fromJWK(jwk: IJwk): KeyPair {
    const data: IJoseData = decodeKeys(jwk);

    return new KeyPair({
      id: jwk.kid,
      algorithms: [jwk.alg as Algorithm],
      created: jwk.created ? new Date(jwk.created) : undefined,
      expires: jwk.expires ? new Date(jwk.expires) : undefined,
      namedCurve: jwk.crv ? (jwk.crv as NamedCurve) : undefined,
      preferredAlgorithm: jwk.alg as Algorithm,
      type: jwk.kty as KeyType,
      ...data,
    });
  }
}
