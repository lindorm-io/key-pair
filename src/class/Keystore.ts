import { JWK } from "../types";
import { KeyPair } from "../entity";
import { KeystoreError } from "../error";
import { filter, find, orderBy, uniqBy } from "lodash";
import { isKeyExpired, isKeyPrivate, isKeyUsable } from "../util";

interface Options {
  keys: Array<KeyPair>;
}

interface TTL {
  seconds: number;
  milliseconds: number;
}

export class Keystore {
  private readonly keys: Array<KeyPair>;

  public constructor(options: Options) {
    const keys = options.keys;

    if (!keys.length) {
      throw new KeystoreError("Keystore was initialised without keys", {
        debug: { keys },
      });
    }

    this.keys = orderBy(keys, ["created", "expires"], ["desc", "asc"]);
  }

  public getJWKS(exposePrivateKeys = false): Array<JWK> {
    const keys: Array<JWK> = [];

    for (const keyPair of this.getKeys()) {
      if (keyPair.external) continue;
      keys.push(keyPair.toJWK(exposePrivateKeys));
    }

    return keys;
  }

  public getKey(id: string): KeyPair {
    const key = find(this.getKeys(), { id });

    if (!key) {
      throw new KeystoreError("Key could not be found", {
        data: { id },
      });
    }

    return key;
  }

  public getKeys(): Array<KeyPair> {
    return filter(uniqBy(this.keys, "id"), isKeyUsable);
  }

  public getPrivateKeys(): Array<KeyPair> {
    return orderBy(filter(this.getKeys(), isKeyPrivate), ["external"], ["asc"]);
  }

  public getSigningKey(): KeyPair {
    const keys = this.getPrivateKeys();

    if (!keys.length) {
      throw new KeystoreError("Private Keys could not be found", {
        debug: { keys: this.keys },
      });
    }

    return keys[0];
  }

  public static getTTL(key: KeyPair): TTL | undefined {
    if (!key.expires) return undefined;
    if (isKeyExpired(key)) return undefined;

    const ttl = key.expires.getTime() - new Date().getTime();

    return {
      seconds: Math.round(ttl / 1000),
      milliseconds: ttl,
    };
  }
}
