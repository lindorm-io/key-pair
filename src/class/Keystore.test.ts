import { Keystore } from "./Keystore";
import { KeyPair } from "../entity";

describe("Keystore.ts", () => {
  let key1: KeyPair;
  let key2: KeyPair;
  let key3: KeyPair;

  let getKeys: () => Array<KeyPair>;

  beforeEach(() => {
    key1 = new KeyPair({
      id: "mock-id-1",
      created: new Date("2020-01-01"),
      expired: null,
      algorithm: "RS512",
      type: "rsa",
      passphrase: "mock-passphrase-1",
      privateKey: "mock-private-key-1",
      publicKey: "mock-public-key-1",
    });
    key2 = new KeyPair({
      id: "mock-id-2",
      created: new Date("2020-02-02"),
      expired: null,
      algorithm: "ES512",
      type: "ec",
      privateKey: "mock-private-key-2",
      publicKey: "mock-public-key-2",
    });
    key3 = new KeyPair({
      id: "mock-id-3",
      created: new Date("2020-03-03"),
      expired: null,
      algorithm: "RS512",
      type: "rsa",
      passphrase: "mock-passphrase-3",
      privateKey: "mock-private-key-3",
      publicKey: "mock-public-key-3",
    });

    getKeys = () => [key1, key2, key3];
  });

  test("should return zero length", () => {
    const store = new Keystore({ keys: undefined });

    expect(store.getLength()).toBe(0);
  });

  test("should return the latest created key", () => {
    const store = new Keystore({ keys: getKeys() });

    expect(store.getCurrentKey().id).toBe("mock-id-3");
  });

  test("should return the latest active key", () => {
    key2 = new KeyPair({
      id: "mock-id-2",
      created: new Date("2020-02-02"),
      expired: new Date("2020-02-03"),
      algorithm: "ES512",
      type: "ec",
      privateKey: "mock-private-key-2",
      publicKey: "mock-public-key-2",
    });
    key3 = new KeyPair({
      id: "mock-id-3",
      created: new Date("2020-03-03"),
      expired: new Date("2020-02-03"),
      algorithm: "RS512",
      type: "rsa",
      passphrase: "mock-passphrase-3",
      privateKey: "mock-private-key-3",
      publicKey: "mock-public-key-3",
    });

    const store = new Keystore({ keys: getKeys() });

    expect(store.getCurrentKey().id).toBe("mock-id-1");
  });

  test("should return a specific key", () => {
    const store = new Keystore({ keys: getKeys() });

    expect(store.getKey("mock-id-2").id).toBe("mock-id-2");
  });

  test("should throw error if no active key can be found", () => {
    key1 = new KeyPair({
      id: "mock-id-1",
      created: new Date("2020-01-01"),
      expired: new Date("2020-02-03"),
      algorithm: "RS512",
      type: "rsa",
      passphrase: "mock-passphrase-1",
      privateKey: "mock-private-key-1",
      publicKey: "mock-public-key-1",
    });
    key2 = new KeyPair({
      id: "mock-id-2",
      created: new Date("2020-02-02"),
      expired: new Date("2020-02-03"),
      algorithm: "ES512",
      type: "ec",
      privateKey: "mock-private-key-2",
      publicKey: "mock-public-key-2",
    });
    key3 = new KeyPair({
      id: "mock-id-3",
      created: new Date("2020-03-03"),
      expired: new Date("2020-02-03"),
      algorithm: "RS512",
      type: "rsa",
      passphrase: "mock-passphrase-3",
      privateKey: "mock-private-key-3",
      publicKey: "mock-public-key-3",
    });

    const store = new Keystore({ keys: getKeys() });

    expect(() => store.getCurrentKey()).toThrowError(new Error("Key could not be found."));
  });
});
