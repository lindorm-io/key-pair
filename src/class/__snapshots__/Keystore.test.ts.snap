// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keystore.ts getJWKS should return keys as private jwks 1`] = `
Array [
  Object {
    "alg": "ES512",
    "created": 1578790800000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": 1640995200000,
    "key_ops": Array [
      "verify",
    ],
    "kid": "publicKeyExpires",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "RS512",
    "created": 1578618000000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "decrypt",
      "verify",
    ],
    "kid": "publicKeyRSA",
    "kty": "RSA",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "ES512",
    "created": 1578445200000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "verify",
    ],
    "kid": "publicKey",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "ES512",
    "created": 1578272400000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": 1640995200000,
    "key_ops": Array [
      "sign",
      "verify",
    ],
    "kid": "privateKeyExpires",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "RS512",
    "created": 1578099600000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "decrypt",
      "encrypt",
      "sign",
      "verify",
    ],
    "kid": "privateKeyRSA",
    "kty": "RSA",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "ES512",
    "created": 1577926800000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "sign",
      "verify",
    ],
    "kid": "privateKey",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
]
`;

exports[`Keystore.ts getJWKS should return keys as public jwks 1`] = `
Array [
  Object {
    "alg": "ES512",
    "created": 1578790800000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": 1640995200000,
    "key_ops": Array [
      "verify",
    ],
    "kid": "publicKeyExpires",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "RS512",
    "created": 1578618000000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "decrypt",
      "verify",
    ],
    "kid": "publicKeyRSA",
    "kty": "RSA",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "ES512",
    "created": 1578445200000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "verify",
    ],
    "kid": "publicKey",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "ES512",
    "created": 1578272400000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": 1640995200000,
    "key_ops": Array [
      "sign",
      "verify",
    ],
    "kid": "privateKeyExpires",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "RS512",
    "created": 1578099600000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "decrypt",
      "encrypt",
      "sign",
      "verify",
    ],
    "kid": "privateKeyRSA",
    "kty": "RSA",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
  Object {
    "alg": "ES512",
    "created": 1577926800000,
    "crv": "crv",
    "d": "d",
    "dp": "dp",
    "dq": "dq",
    "e": "e",
    "expires": undefined,
    "key_ops": Array [
      "sign",
      "verify",
    ],
    "kid": "privateKey",
    "kty": "EC",
    "n": "n",
    "p": "p",
    "q": "q",
    "qi": "qi",
    "use": "sig",
    "x": "x",
    "y": "y",
  },
]
`;
